Reduce coupling in Population and GA now takes initial solutions

leonlan opened the pull request:
This PR:
- Population no longer depends on problem data, penalty manager and Xorshift.
- Closes #71: initial population solutions are no longer initialised in Population itself.
- GeneticAlgorithm now takes as `initial_solutions` argument, which is required to be non-empty.

leonlan commented:
@N-Wouda This is a proposal to close #71, to address the discussion in #156, and to address coupling in Population #157. I have not yet changed tests and notebooks, which I'll do if we decide to proceed with these changes.

leonlan commented on pyvrp/Population.py:
```diff
@@ -15,38 +13,34 @@ class Population:
     ):
-        self._data = data
-        self._pm = penalty_manager
         self._rng = rng
```
We can also remove the RNG dependency and pass it as argument to `self.select`/`self.binary_tournament`. I don't know if that's desirable.

N-Wouda replied:
I think it's better because there's less internal state on `Population` then.

N-Wouda commented on pyvrp/Population.py:
```diff
@@ -138,14 +132,14 @@ def select(self) -> Tuple[Individual, Individual]:
         """
-        Restarts the population. All individuals are removed and a new initial
-        population population is generated.
+        Restarts the population. All current individuals are removed and the
+        initial individuals used to seed the restarted population.
         """
         self._feas = SubPopulation(self._op, self._params)
         self._infeas = SubPopulation(self._op, self._params)
 
-        for _ in range(self._params.min_pop_size):
-            self.add(Individual.make_random(self._data, self._pm, self._rng))
+        for indiv in self._initial_solutions:
+            self.add(indiv)
```
Do you know if this impacts performance? I don't really have a good intuition for this.

N-Wouda replied:
I'd assume no, but it might not hurt to make sure.

leonlan replied:
I'll run a benchmark just to be sure.

leonlan commented:
Benchmark: GH1000, 3600s, 1 seed

[3f14e09](https://github.com/N-Wouda/PyVRP/pull/168/commits/3f14e09a63bf8cd7cfcc5070e344a3bba8ac6b1c)
```
Cost:  333780.9
Iters: 102301.9
```

main
```
TODO
```

wouterkool commented:
Shouldn't we use a different set of initial solutions when restarting?

leonlan commented:
> Shouldn't we use a different set of initial solutions when restarting?

I don't think it matters. The binary selection, crossovers, and local search introduce randomness. So even if we restart with the same initial population, the resulting trajectory will be different each time.

leonlan commented:
This PR is ready to be reviewed. Meanwhile, I'll run a 10 seed benchmark to check that there is no performance drop with the new change (restart with the same individuals).

N-Wouda commented:
I'll try to have a look at this tonight.

N-Wouda commented on pyvrp/Population.py:
```diff
@@ -131,26 +126,32 @@ def select(self) -> Tuple[Individual, Individual]:
         """
-        Restarts the population. All individuals are removed and a new initial
-        population population is generated.
+        Restarts the population. All current individuals are removed and the
+        original initial individuals are used agains to initialize the
+        restarted population.
```
Typo: agains.

Can we rewrite this a bit to really clarify that we're restarting from the initial individuals that were passed when this object was initialised? It's not clear enough to me what "original initial individuals" means.

N-Wouda commented on pyvrp/Population.py:
```diff
@@ -15,12 +13,8 @@ class Population:
-    rng
-        Random number generator.
+    initial_solutions
+        List of individuals used to initialize the population.
```
```suggestion
        List of individuals used to initialise the population.
```

N-Wouda commented on pyvrp/Population.py:
```diff
@@ -30,23 +24,19 @@ class Population:
 
-        for _ in range(params.min_pop_size):
-            self.add(Individual.make_random(data, penalty_manager, rng))
+        for indiv in initial_solutions:
+            self.add(indiv)
```
```suggestion
        for individual in initial_solutions:
            self.add(individual)
```

N-Wouda commented on pyvrp/Population.py:
```diff
@@ -131,26 +126,32 @@ def select(self) -> Tuple[Individual, Individual]:
 
-        for _ in range(self._params.min_pop_size):
-            self.add(Individual.make_random(self._data, self._pm, self._rng))
+        for indiv in self._initial_solutions:
+            self.add(indiv)
```
```suggestion
        for individual in self._initial_solutions:
            self.add(individual)
```

N-Wouda commented on pyvrp/plotting/tests/test_plotting.py:
```diff
@@ -52,9 +59,13 @@ def test_plot_result():
 
     data = read("data/RC208.txt", "solomon", round_func="trunc")
     bks = read_solution("data/RC208.sol")
+
     pm = PenaltyManager(data.vehicle_capacity)
     rng = XorShift128(seed=42)
-    pop = Population(data, pm, rng, broken_pairs_distance)
+    params = PopulationParams()
+    init = make_random_solutions(params.min_pop_size, pm, rng, data)
+
+    pop = Population(init, broken_pairs_distance, params=params)
     stats = Statistics()
```
The spacing here doesn't make much sense to me. I'd expect `params`, `init`, and `pop` to all be together in the same paragraph (since they're all part of "population stuff").

N-Wouda commented on pyvrp/tests/test_Population.py:
```diff
@@ -257,12 +277,13 @@ def test_elite_individuals_are_not_purged(nb_elite: int):
 
 def test_binary_tournament_ranks_by_fitness():
-    data = read("data/RC208.txt", "solomon", "dimacs")
+    data = read("data/RC208LessVehicles.txt", "solomon", "dimacs")
```
Why does this need a different (new) instance?

leonlan replied:
Because the test assumes that the randomly generated individuals are all infeasible, which was not the case when I ran the test with the original file.

N-Wouda replied:
But we can just filter the feasible solutions away before passing `init` to `pop`, right?

N-Wouda commented on pyvrp/tests/data/RC208LessVehicles.txt:
```diff
@@ -0,0 +1,110 @@
+RC208
```
If we can, I'd like to avoid introducing new test data.

N-Wouda commented on pyvrp/tests/test_Population.py:
```diff
@@ -177,24 +178,43 @@ def test_select_returns_same_parents_if_no_other_option():
+    # passed-in initial solutions. We check for equality here because the
+    # population creates new individuals.
+    for indiv in pop:
+        assert_(np.any(indiv == other for other in init))
 
-    old = {individual for individual in pop}
     pop.restart()
-    new = {individual for individual in pop}
 
-    assert_equal(len(pop), params.min_pop_size)
-    assert_equal(len(old & new), 0)  # no old pops survived the restart
+    # Check again for the restarted population.
+    for indiv in pop:
+        assert_(np.any(indiv == other for other in init))
```
I don't really get why this needs to be changed. Would something like:
```python
init = set(...)
current = {individual for individual in pop}
assert_(len(current & init), params.min_pop_size)

pop.restart()

new = {individual for individual in pop}
assert_(len(new & init), params.min_pop_size)
```
not work?

leonlan replied:
No idea why I changed this - I changed it again to what you suggested!

N-Wouda commented on pyvrp/tests/test_Population.py:
```diff
@@ -177,24 +178,43 @@ def test_select_returns_same_parents_if_no_other_option():
+
+@mark.parametrize("num_init", [1, 15, 21, 30, 100])
+def test_num_initial_solutions(num_init):
+    data = read("data/OkSmallFeasible.txt")
```
Do we need a new test data instance for this?

N-Wouda commented on pyvrp/tests/helpers.py:
```diff
@@ -37,6 +38,13 @@ def sleep(duration, get_now=time.perf_counter):
         now = get_now()
 
 
+def make_random_solutions(num_sols, pm, rng, data):
+    """
+    Returns a list of a ``num_sols`` random solutions.
+    """
+    return [Individual.make_random(data, pm, rng) for _ in range(num_sols)]
```
Can we have the order of the arguments here be the same as they are for almost every other thing: `data, pm, rng`? I'm fine with `num_sols` coming first.

N-Wouda commented on pyvrp/Population.py:
```diff
@@ -30,23 +24,19 @@ class Population:
-        data: ProblemData,
-        penalty_manager: PenaltyManager,
-        rng: XorShift128,
+        initial_solutions: List[Individual],
```
Can we make the type here a general iterable? It doesn't necessarily have to be a list.

N-Wouda replied:
I think I'd make sense for `initial_solutions` to be optional. By default we construct an empty population. This'd be possible if we did, for example,
```python
def __init__(
  self,
  diversity_op,
  initial_solutions: Iterable[Individual] = tuple(),
  params: PopulationParams = PopulationParams()
)
```

leonlan replied:
I don't think that making the initial solution argument is desirable. Creating an empty Population is not a good default, and it also doesn't work well with the restarting mechanism.

N-Wouda replied:
>Creating an empty Population is not a good default

Why not? That Population should not be empty is a problem for the GeneticAlgorithm, not for the Population container itself.

>and it also doesn't work well with the restarting mechanism.

This I don't understand. Restart would just return to an empty population again; I don't see what'd go wrong.

N-Wouda replied:
Thinking about this, we should probably enforce in the GA that the population has least 1/2 (I think 2?) individuals, right now it can be empty.

N-Wouda replied:
I'll add a PR for that.

N-Wouda commented on pyvrp/tests/data/OkSmallFeasible.txt:
```diff
@@ -0,0 +1,41 @@
+NAME : TEST
```
If we can, I'd like to avoid introducing new test data.

N-Wouda commented on pyvrp/tests/helpers.py:
```diff
@@ -37,6 +38,13 @@ def sleep(duration, get_now=time.perf_counter):
 
+def make_random_solutions(num_sols, pm, rng, data):
+    """
+    Returns a list of a ``num_sols`` random solutions.
```
```suggestion
    Returns a list of ``num_sols`` random solutions.
```

N-Wouda commented on pyvrp/tests/test_Population.py:
```diff
@@ -177,24 +178,43 @@ def test_select_returns_same_parents_if_no_other_option():
+    init = make_random_solutions(num_init, pm, rng, data)
+    pop = Population(init, bpd, params)
+
+    # If there are more initial individuals than the maximal population size
+    # allows, then a purge is triggered. This resulting population size is
+    # equal to ``num_init`` modulo the number individuals that need to
+    # be added to trigger a purge (11).
+    assert_equal(len(pop), num_init % 11)
```
This isn't true in general. Suppse we add `15` individuals. Then `15 % 11 == 4`. But what if we've added 7 feasible and 8 infeasible solutions? Then no purge would have been triggered because neither subpopulation would have been of maximal size.

The reason this test passes is that all solutions are feasible in this instance, which is not explicitly mentioned in the test. Can we document that much more explicitly?

N-Wouda commented on pyvrp/tests/test_GeneticAlgorithm.py:
```diff
@@ -92,17 +93,21 @@ def test_params_constructor_does_not_raise_when_arguments_valid(
+    init = make_random_solutions(pop_params.min_pop_size, pm, rng, data)
+    pop = Population(init, bpd, params=pop_params)
+
+    ls = LocalSearch(data, pm, rng, compute_neighbours(data))
+
+    node_op = Exchange10(data, pm)
     ls.add_node_operator(node_op)
```
While we're here anyway:

```suggestion
    ls = LocalSearch(data, pm, rng, compute_neighbours(data))
    ls.add_node_operator(Exchange10(data, pm))
```

N-Wouda commented on pyvrp/tests/test_Population.py:
```diff
@@ -143,15 +144,15 @@ def test_select_returns_same_parents_if_no_other_option():
 
     params = PopulationParams(min_pop_size=0)
-    pop = Population(data, pm, rng, broken_pairs_distance, params)
+    pop = Population([], bpd, params=params)
```
I was thinking about whether we should make `params` a keyword only argument. Perhaps yes, but then not in this PR (since we should then probably enforce it all at once for every Params-like object).

N-Wouda reviewed (commented):
A bunch of small things and some questions. On the whole this looks good to me!

leonlan commented:
I'll address the remaining comments tomorrow morning.

N-Wouda commented on pyvrp/GeneticAlgorithm.py:
```diff
@@ -73,20 +75,27 @@ def __init__(
-        self._best = Individual.make_random(data, penalty_manager, rng)
+        # Initialise best using the feasible initial solution with lowest cost,
+        # or the initial solution with lowest cost if none are feasible.
+        self._best = min(
+            self._initial_solutions,
+            key=lambda indiv: (not indiv.is_feasible(), indiv.cost()),
+        )
```
Here, and in a test: this is a bit too clever IMHO. I had to think for too long about why this works, so its readability could be improved.

Something like
```python
# Find best feasible initial solution if any exist, else set lower cost
# solution as initial best.
init_feas = [indiv for indiv in initial_solutions if indiv.is_feasible()]
sols = init_feas if init_feas else initial_solutions
self._best = min(sols, key=lambda indiv: indiv.cost())
```
is barely longer, but doesn't make me think as hard.

In the test I think we can add some feasible individual that we know is the best, and then test whether that individual is indeed set as the best. There's no need to reproduce the logic there.

N-Wouda reviewed (approved):
LGTM! A few small things about how these changes can be made a bit more legible, but nothing that's blocking merging this in.

leonlan commented:
> LGTM! A few small things about how these changes can be made a bit more legible, but nothing that's blocking merging this in.

OK, I addressed all comments. Thanks for all the help! I'll run a benchmark before merging this PR.

leonlan commented:
GH1000, 10 seeds, 36000s
```
min     33341.5
mean    33359.6
max     33374.5
```
All looks fine, so I'm merging this in.
